pkg/controllers/pluginconfig/pluginbundle_controller.go

```diff
@@ -0,0 +1,89 @@
+
+package pluginconfig
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	greenhousev1alpha1 "github.com/cloudoperators/greenhouse/pkg/apis/greenhouse/v1alpha1"
+	"github.com/cloudoperators/greenhouse/pkg/clientutil"
+)
+
+// PluginBundleReconciler reconciles a PluginBundle object
+type PluginBundleReconciler struct {
+	client.Client
+	recorder record.EventRecorder
+}
+
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=pluginbundles,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=pluginbundles/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=pluginbundles/finalizers,verbs=update
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=pluginconfigs,verbs=get;list;watch;create;update;patch;delete
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PluginBundleReconciler) SetupWithManager(name string, mgr ctrl.Manager) error {
+	r.Client = mgr.GetClient()
+	r.recorder = mgr.GetEventRecorderFor(name)
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&greenhousev1alpha1.PluginBundle{}).
+		Complete(r)
+}
+
+func (r *PluginBundleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	var pluginBundle = new(greenhousev1alpha1.PluginBundle)
+	if err := r.Get(ctx, req.NamespacedName, pluginBundle); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if err := r.reconcilePluginBundle(ctx, pluginBundle); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *PluginBundleReconciler) reconcilePluginBundle(ctx context.Context, pluginBundle *greenhousev1alpha1.PluginBundle) error {
+	lblSelector, err := v1.LabelSelectorAsSelector(&pluginBundle.Spec.ClusterSelector)
+	if err != nil {
+		return err
+	}
+	var selectedClusters = new(greenhousev1alpha1.ClusterList)
+	if err := r.List(ctx, selectedClusters, client.InNamespace(pluginBundle.GetNamespace()), client.MatchingLabelsSelector{Selector: lblSelector}); err != nil {
+		return err
+	}
+
+	var allErrs = make([]error, 0)
+	for _, pluginConfigSpec := range pluginBundle.Spec.Plugins {
+		for _, cluster := range selectedClusters.Items {
+			var pluginConfigForCluster = new(greenhousev1alpha1.PluginConfig)
+			pluginConfigForCluster.Namespace = pluginBundle.GetNamespace()
+			pluginConfigForCluster.Name = makeNameForPluginConfigFromPluginBundle(&pluginConfigSpec, &cluster)
+			result, err := clientutil.CreateOrPatch(ctx, r.Client, pluginConfigForCluster, func() error {
+				pluginConfigForCluster.Spec = pluginConfigSpec
+				pluginConfigForCluster.Spec.ClusterName = cluster.GetName()
+				return nil
+			})
```

Comment:
In the case that an PluginAdmin changes the `OptionValues` for the generated `Plugin` it might not be obvious why the values are overwritten from time to time.
I can image three solutions with increasing effort:

- Only create _Plugins_ if they are non-existing
- Make it obvious to the enduser where they need to perform changes to ensure the changes are persistent
- Annotate the `Plugins` created by this controller. This way the annotation can be removed if the Plugin shall no longer be managed by this controller

Reply:
As discussed in the weekly meeting: Updates to Plugins that have been created by the PluginPreset controller will be forbidden. This will be enforced by the AdmissionWebhooks and the UI by checking for a `OwnerReference` on the _Plugin_ pointing to a _PluginPreset_

Reply:
Edits can't be forbidden, but will return a `Warning` and the PluginPresetController will reconcile the PluginPreset for which the dependent Plugin was updated.